### PR TITLE
Update template to use statici18n

### PR DESCRIPTION
**Pull request by viadanna**
This PR contains modifications to the template to make use of `django-statici18n` to provide translations of static resources. It generates a static javascript that provides `gettext` to be used in static resources without the need to ping the server for translations.

The XBlock is loaded as an INSTALLED_APP in Django to make it possible for `django_statici18n` to generate the translation javascript.

**Merge deadline**: None

**Testing instructions**:

1. Generate an XBlock using this template.
2. Load the generated XBlock in xblock-sdk using `make`
3. Navigate to the XBlock and make sure `test.js` for the selected language loads correctly.

**Author notes and concerns**:

1. This is using a forked repo of `django-statici18n` while its author reviews the pull request needed to make this work. Will update to a release when accepted.

**Reviewers**
- [x] @rocioar
- [ ] edX reviewer[s] TBD

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @viadanna! I've created [OSPR-2421](https://openedx.atlassian.net/browse/OSPR-2421) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/cookiecutter-xblock/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

File: {{cookiecutter.repo_name}}/manage.py
Comment (rocioar) [inline {{cookiecutter.repo_name}}/manage.py]: nit: could you move the parenthesis one line down? so that it looks like this:

```
os.environ.setdefault(
    "DJANGO_SETTINGS_MODULE",
    "{{cookiecutter.package_name}}.translations.settings"
)
```

File: {{cookiecutter.repo_name}}/manage.py
Comment (rocioar) [inline {{cookiecutter.repo_name}}/manage.py]: nit: what do you think of moving this to the top? I think it would be easier to read the imports if they are in the top of the file

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py
Comment (rocioar) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py]: nit: what do you think of doing:

```
if not lang_code:
    return
```

I think it's clearer to read.

**Review by rocioar — COMMENTED** [review]
@viadanna just left some minor comments, nitpicks mostly. Other than that the code looks good, I will test this and update here.

**Comment by rocioar** [comment]
@viadanna I've tested this and it's working fine. Please review the nits. Good work!

File: {{cookiecutter.repo_name}}/manage.py
Comment (viadanna) [reply {{cookiecutter.repo_name}}/manage.py]: 👍

File: {{cookiecutter.repo_name}}/manage.py
Comment (viadanna) [reply {{cookiecutter.repo_name}}/manage.py]: Oh, forgot about that 👍

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py
Comment (viadanna) [reply {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py]: Sure, it does improve readability.

**Comment by mduboseedx** [comment]
@viadanna Thank you for the contribution! I am curious as to how you are using this repo. We actually haven't managed this repo for some time at edX as it is missing some of our normal configuration files that we use for OSPR automation.

Are you planning to do a number of changes on this repo?

**Comment by viadanna** [comment]
@mduboseedx Actually, this change was only meant to include compiled i18n of static resources, I did not expect to include other changes. Is it missing anything?

**Comment by mduboseedx** [comment]
@viadanna I don't believe that there is anything extra you'll need to do on your end, but let me check with my team on whether we need to do anything with this repo. Thank you for your patience!

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/static/js/src/{{cookiecutter.package_name}}.js
Comment (bryanlandia) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/static/js/src/{{cookiecutter.package_name}}.js]: @viadanna looks like a backwards `/*` here

Comment (viadanna) [reply {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/static/js/src/{{cookiecutter.package_name}}.js]: Yeah, let me correct that.

**Comment by nedbat** [comment]
@edx/platform-team @jmbowman Who should own this?

File: {{cookiecutter.repo_name}}/requirements.txt
Comment (macdiesel) [inline {{cookiecutter.repo_name}}/requirements.txt]: Currently the edx-platform is running version 1.4, I think we would want to match this version.

Comment (viadanna) [reply {{cookiecutter.repo_name}}/requirements.txt]: Hi @macdiesel 

It requires this version of `django-statici18n` as it contains changes to allow for namespacing the JS translation tooling.

Comment (macdiesel) [reply {{cookiecutter.repo_name}}/requirements.txt]: So we would need to upgrade edx-platform and other repos using django-statici18n to 1.7 before I could accept this PR.

Comment (viadanna) [reply {{cookiecutter.repo_name}}/requirements.txt]: Thanks @macdiesel 

I'll just send another commit to use the upstream version now that our changes were merged.

File: {{cookiecutter.repo_name}}/Makefile
Comment (iamjazzar) [inline {{cookiecutter.repo_name}}/Makefile:15]: 👍

**Review by iamjazzar — APPROVED** [review]
I've tested this and it LGTM 👍

File: {{cookiecutter.repo_name}}/Makefile
Comment (iamjazzar) [inline {{cookiecutter.repo_name}}/Makefile]: 👍

**Comment by OmarIthawi** [comment]
@viadanna thanks for the improvement! I see that we're having two resources, one for Python and the other for JavaScript, is that correct?

Would it makes sense to add `.tx/config` with this PR, or do you feel that's it's out of scope?

File: {{cookiecutter.repo_name}}/Makefile
Comment (OmarIthawi) [inline {{cookiecutter.repo_name}}/Makefile]: `tx push ...ru,zh_CN` -> `i18n_tool transifex push`

The latter will read `config.yaml` and push the needed languages.

Comment (OmarIthawi) [inline {{cookiecutter.repo_name}}/Makefile]: Do you mind explaining why would you need to push the translations? Usually it's only the source language.

It's also very beneficial to add `extract_translations` as dependency.

Please amend to the following:
```
push_translations: extract_translations ## push translations to transifex
	i18n_tool transifex push
```

File: {{cookiecutter.repo_name}}/Makefile
Comment (OmarIthawi) [inline {{cookiecutter.repo_name}}/Makefile]: `tx pull -f --mode=reviewed -l en,ar,es_419,fr,he,hi,ko_KR,pt_BR,ru,zh_CN` -> `i18n_tool transifex pull`

As long as `config.yaml` is up to date, it will work just fine, and keep the langs in a single place.

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/locale/config.yaml
Comment (OmarIthawi) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/locale/config.yaml:92]: Please add `public/js/translations` to the ignored dirs. Otherwise `i18n_tool extract` (and `django-admin`) will exit with cryptic errors.

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py
Comment (OmarIthawi) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py]: `lang_code` -> `locale_code`

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py
Comment (OmarIthawi) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py]: Is that really possible, to have `lang_code` as falsy?

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py
Comment (OmarIthawi) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py]: The logic is correct but the variable name is misleading here. As country code can't be a locale.

`lang_code, _country_code = lang_code.split('-')`

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py
Comment (OmarIthawi) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py]: Subsequently: `for code in (lang_code, country_code, 'en'):` -> `for code in (locale_code, lang_code, 'en'):`

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/locale/README.txt
Comment (OmarIthawi) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/locale/README.txt]: `compile_jsi18n.sh` is probably now baked into Makefile, right?

File: {{cookiecutter.repo_name}}/Makefile
Comment (OmarIthawi) [inline {{cookiecutter.repo_name}}/Makefile]: `&& DJANGO_SETTINGS_MODULE="$(SETTINGS_FILE)"` isn't needed here AFAIK.

File: {{cookiecutter.repo_name}}/Makefile
Comment (OmarIthawi) [inline {{cookiecutter.repo_name}}/Makefile]: This should break since Django will try to import `locale` and that will cause confusion. AFAIK it will only work if it's moved to `$(PACKAGE_NAME).settings`. The actual file should be renamed as well.

File: {{cookiecutter.repo_name}}/manage.py
Comment (OmarIthawi) [inline {{cookiecutter.repo_name}}/manage.py:1]: It seems that this file isn't used, if so please remove it.

Unless we have to use to for `runserver`, which is AFAIK not the case.

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/locale/config.yaml
Comment (OmarIthawi) [inline {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/locale/config.yaml:1]: The `i18n_service` expects this directory to be named `translations` (the support that Jill introduced):

https://github.com/edx/xblock-utils/blob/18fcd70ff8eed6af17d41b16cf66a8161c9fb51b/tests/unit/test_resources.py#L132-L148

**Review by OmarIthawi — CHANGES_REQUESTED** [review]
Thanks @viadanna. This PR should save a lot of mundane work.

I've added some comments. I highly appreciate it if you can address them.

File: {{cookiecutter.repo_name}}/Makefile
Comment (viadanna) [reply {{cookiecutter.repo_name}}/Makefile]: Hey @OmarIthawi thanks a lot! This Makefile is still WIP so feel free to make suggestions here.

Comment (viadanna) [reply {{cookiecutter.repo_name}}/Makefile]: No need at all, just testing the transifex service.

Good idea extracting it before pushing.

File: {{cookiecutter.repo_name}}/Makefile
Comment (viadanna) [reply {{cookiecutter.repo_name}}/Makefile]: That's what I was looking for 👍

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py
Comment (viadanna) [reply {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py]: Alright

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py
Comment (viadanna) [reply {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py]: Not False, but [None](https://docs.djangoproject.com/en/dev/ref/utils/#django.utils.translation.get_language) is possible. Better use `is None` instead.

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py
Comment (viadanna) [reply {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py]: Are we sure we'll always get a [language]-[country] here?

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py
Comment (viadanna) [reply {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py]: 👍

File: {{cookiecutter.repo_name}}/Makefile
Comment (viadanna) [reply {{cookiecutter.repo_name}}/Makefile]: 👍

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/locale/README.txt
Comment (viadanna) [reply {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/locale/README.txt]: Yeah

File: {{cookiecutter.repo_name}}/Makefile
Comment (viadanna) [reply {{cookiecutter.repo_name}}/Makefile]: Yeah, this just works with the xblock installed to the virtualenv. I'm moving back to the `manage.py` to get rid of this confusion.

File: {{cookiecutter.repo_name}}/manage.py
Comment (viadanna) [reply {{cookiecutter.repo_name}}/manage.py:1]: As said before, I'm moving back to this to fix the issues mentioned.

File: {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/locale/config.yaml
Comment (viadanna) [reply {{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/locale/config.yaml:1]: Alright, I'll add a symlink as `locale` is also necessary.

File: {{cookiecutter.repo_name}}/TRANSLATING.txt
Comment (OmarIthawi) [inline {{cookiecutter.repo_name}}/TRANSLATING.txt]: Same here please `compile_jsi18n.sh` isn't there anymore :).

**Review by OmarIthawi — COMMENTED** [review]
Thanks @viadanna for addressing the review comments. I would like to recommend exercising this PR against Travis (see PR https://github.com/edx/cookiecutter-xblock/pull/5). Please feel free to go without Travis if you think that it's out of the scope of this PR.

TL;DR;
I've passed through each comment and marked the ones that have been addressed as "Resolved" (Using GitHub's new [resolution button](https://help.github.com/articles/commenting-on-a-pull-request/#resolving-conversations)). 

There are other minor points that haven't been addressed but shouldn't block the PR.

**Review by OmarIthawi — APPROVED** [review]
It seems that I can't dismiss my "request changes" review, without marking it as "approved".

This PR adds a lot of value, we should get it merged even if it has a couple outstanding minor issue.

Thanks again @viadanna!

File: {{cookiecutter.repo_name}}/TRANSLATING.txt
Comment (viadanna) [reply {{cookiecutter.repo_name}}/TRANSLATING.txt]: Wrong commit, I actually meant to merge these with README.rst

**Comment by bmedx** [comment]
Hi @viadanna the changes look good so far, but the instructions on the main README.rst no longer work. I was able to run my xblock by using `make dev.run` instead of `docker run`, but the `docker run` line could also just be updated to match what's in the Makefile. If you can get that updated I think we'll be ready to merge this!

**Comment by viadanna** [comment]
I've updated the README and squashed the commits.

Thanks a lot @bmedx

File: README.rst
Comment (bmedx) [inline README.rst]: This line still didn't work for me when using the cookiecutter defaults. This did:

`docker run -p 8000:8000 -v $(pwd):/usr/local/src/myxblock-xblock --name myxblock-xblock-dev myxblock-xblock-dev`

And `make dev.run` did, which upon further consideration I think is probably better to suggest to users. It does image cleanup and is a lot easier to remember. :)

Comment (viadanna) [reply README.rst]: That's odd, I tested it working before committing. Instructing to use `make dev.run` makes more sense.

**Comment by openedx-webhooks** [comment]
@viadanna 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/cookiecutter-xblock&entry.867055334=https://github.com/edx/cookiecutter-xblock/pull/4&entry.1484655318=https://github.com/viadanna&entry.752974735=2018-05-17+15:58&entry.1917517419=2018-09-26+13:48&entry.2133058324=Yes) so we can improve your experience in the future.

**Comment by bmedx** [comment]
@viadanna all set, thank you for the contribution!

**Comment by viadanna** [comment]
Alright, thanks a lot @bmedx !

**Pull request merged**
